rfc_samplers.md

```diff
@@ -0,0 +1,125 @@
+
+Feature space
+-------------
+
+At a high level, a sampler returns a sequence of "clips" where a clip is a batch
+of (non-necessarily contiguous) frames. Based on an analysis of the sampling
+options of other libraries (torchvision, torchmultimodal, and various internal
+libraries), we want to support the following use-cases:
+
+
+**Clip location** (and number thereof)
+
+- Get `num_clips_per_video` random clips: typically called "Random" sampler
+- Get `num_clips_per_video` equally-spaced clips: "Uniform" sampler. Space is
+  index-based.
+- Get `num_clips_per_second` clips per seconds [within the [`start_seconds`,
+  `stop_seconds`] sampling range]: "Periodic" sampler.
+
+**Clip content**
+
+- Clip must be `num_frames_per_clip` frames spaced by `dilation` frames.
+
+Notes:
+
+- There are 3 sampling strategies defining the clip locations: Random, Uniform
+  and Periodic. The names are prone to confusion. "Random" is in fact uniformly
+  random while "Uniform" isn't random. "Uniform" is periodic and "Periodic" is
+  uniformly spaced too.
+- The parametrization of the *Clip Content* dimension is limited and
+  reduces to just `num_frames_per_clip` and the index-based `dilation`.
+- The *Clip Content* dimension is never random. `num_frames_per_clip` is always
+  passed by the user. This makes sense: training typically requires fixed-sized
+  inputs so `num_frames_per_clip` is expected to be constant. The content of a
+  clip (space between frames) is always regular, never random.
+
+
+Proposal
+--------
+
+```py
+from torchcodec.decoders import SimpleVideoDecoder
+from torchcodec import samplers
+
+decoder = SimpleVideoDecoder("./cool_vid")
+
+
+def samplers.clips_at_random_timestamps(
+    decoder,
+    *,
+    num_clips,
+    num_frames_per_clip=1,
+    seconds_between_frames=None,
+) -> FrameBatch
+
+def samplers.clips_at_regular_indices(
+    decoder,
+    *,
+    num_clips,
+    num_frames_per_clip=1,
+    num_indices_between_frames=1,  # in indices (dilation)
+) -> FrameBatch
+
+def samplers.clips_at_regular_timestamps(
+    decoder,
+    *,
+    num_clips_per_second,
```

Comment:
I find `num_clips_per_second` confusing to reason about because we need to think about fractions of a clip. That is, because we've fixed the unit to be 1 second ("per second"), if we want a clip to span multiple seconds, we'll need to specify what fraction of the clip is taken up by 1 second. For example, if a user wants a clip to start every 3 seconds, I believe they would need to say `num_clips_per_second=1/3`. That is, each second has 1/3 of a clip, so we need 3 seconds before we get a full clip.

I have been thinking of how to specify the start in time as a period instead of a rate. So more intuitive to me is `clips_start_period_seconds`. In order to achieve the above, I would say `clips_start_period_seconds=3`. A clip starts every 3 seconds. Granted, if I want multiple clips in one second, we have to provide values less than 1 again: `clips_start_period_seconds=1/3` would be starting a clip every 1/3 of a second.

Personally, I find it more intuitive to think of fractions of a second than fractions of a clip. And I find it more intuitive to think in terms of a period instead of a rate.

Reply:
I think it would be more consistent if the user specifies seconds_between_clip_starts and seconds_between_samples (or seconds_between_frames) (or vice versa).

Reply:
`num_clips_per_seconds` has 2 clear advantages over `clips_start_period_seconds` IMHO:

- It is what existing samplers use. It reduces migration friction.
- It is immediately obvious what it means, while that's not the case for `clips_start_period_seconds`

> Personally, I find it more intuitive to think of fractions of a second than fractions of a clip

I share the same view, and that's precisely why I tend to think `num_clips_per_seconds` is easier to grasp: to me, a fraction of a second implies a `per_second` parameter.

Note that I don't think this premise is quite exact: 

> if we want a clip to span multiple seconds, we'll need to specify what fraction of the clip is taken up by 1 second

The time spanned by a clip isn't just determined by this parameter, it also depend on the number of frames within that clip (and the framerate). 

Happy to chat more!

Reply:
FWIW, I like `seconds_between_clip_starts` and `seconds_between_samples_within_clip` (both are periods, not frequencies).